File: jl/reservoir_computing/src/ESN.jl
module ESN

using LinearAlgebra
using Random
using Statistics

export ESNConfig, EchoStateNetwork, build_esn, train_esn!, predict_esn

Base.@kwdef struct ESNConfig
    N_reservoir::Int    = 100
    sparsity::Float64   = 0.1
    spectral_radius::Float64 = 0.9
    input_scaling::Float64   = 0.5
    leak_rate::Float64       = 1.0
    ridge_alpha::Float64     = 1e-6
    seed::Int           = 42
end

mutable struct EchoStateNetwork
    cfg::ESNConfig
    W_in::Matrix{Float64}    # N_res × n_in
    W_res::Matrix{Float64}   # N_res × N_res
    W_out::Union{Matrix{Float64}, Nothing}  # n_out × (N_res + n_in)
    state::Vector{Float64}   # current reservoir state
    last_input::Vector{Float64}  # last training input (for autonomous prediction)
end

function build_esn(cfg::ESNConfig; n_input::Int = 3)
    rng = MersenneTwister(cfg.seed)
    N   = cfg.N_reservoir

    W_in  = (2rand(rng, N, n_input) .- 1) .* cfg.input_scaling

    # Sparse random reservoir. Sample distinct entries without replacement so the
    # realised connection density equals `sparsity` exactly — the sweep axis of
    # the phase diagram. (Sampling with replacement loses ~5% of edges to
    # collisions, biasing the very quantity being measured.)
    W_res = zeros(Float64, N, N)
    n_edges = round(Int, N * N * cfg.sparsity)
    for k in randperm(rng, N * N)[1:n_edges]
        W_res[k] = randn(rng)
    end
    # Rescale to desired spectral radius
    ρ = maximum(abs.(eigvals(W_res)))
    ρ > 1e-8 && (W_res .*= cfg.spectral_radius / ρ)

    EchoStateNetwork(cfg, W_in, W_res, nothing, zeros(N), zeros(n_input))
end

function _reservoir_step!(esn::EchoStateNetwork, u::Vector{Float64})
    cfg = esn.cfg
    pre = esn.W_res * esn.state .+ esn.W_in * u
    new_state = tanh.(pre)
    esn.state .= (1 - cfg.leak_rate) .* esn.state .+ cfg.leak_rate .* new_state
    esn.state
end

function train_esn!(esn::EchoStateNetwork, data::Matrix{Float64}, washout::Int)
    n, d = size(data)
    N    = esn.cfg.N_reservoir
    esn.state .= 0.0

    # Collect reservoir states
    states_collected = zeros(Float64, n - washout - 1, N + d)
    for t in 1:n-1
        _reservoir_step!(esn, data[t, :])
        if t > washout
            states_collected[t - washout, 1:N] .= esn.state
            states_collected[t - washout, N+1:end] .= data[t, :]
        end
    end

    targets = data[washout+2:end, :]
    alpha   = esn.cfg.ridge_alpha
    A = states_collected' * states_collected + alpha * I(N + d)
    B = states_collected' * targets
    esn.W_out = (A \ B)'
    # Advance one more step so state matches end-of-training
    _reservoir_step!(esn, data[end, :])
    # Save last training input for prediction initialisation
    esn.last_input = copy(data[end, :])
    nothing
end

function predict_esn(esn::EchoStateNetwork, n_steps::Int)
    isnothing(esn.W_out) && error("ESN not trained. Call train_esn! first.")
    N = esn.cfg.N_reservoir
    d = size(esn.W_out, 1)
    preds = zeros(n_steps, d)
    inp   = copy(esn.last_input)   # seed from last training point
    for t in 1:n_steps
        _reservoir_step!(esn, inp)
        extended = vcat(esn.state, inp)
        out = esn.W_out * extended
        preds[t, :] .= out
        inp .= out
    end
    preds
end

end

File: py/reservoir_computing/esn.py
"""Echo State Network implementation."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

import numpy as np

Array = np.ndarray


@dataclass
class ESNConfig:
    N_reservoir: int = 100
    sparsity: float = 0.1
    spectral_radius: float = 0.9
    input_scaling: float = 0.5
    leak_rate: float = 1.0
    ridge_alpha: float = 1e-6
    seed: int = 42


class EchoStateNetwork:
    def __init__(self, cfg: ESNConfig, n_input: int = 3):
        self.cfg = cfg
        rng = np.random.default_rng(cfg.seed)
        N = cfg.N_reservoir
        self.W_in = (2*rng.random((N, n_input)) - 1) * cfg.input_scaling
        # Sample distinct entries without replacement so the realised connection
        # density equals `sparsity` exactly — the sweep axis of the phase diagram.
        # (Sampling with replacement loses ~5% of edges to collisions.)
        W_res = np.zeros(N * N)
        n_edges = round(N * N * cfg.sparsity)
        idx = rng.choice(N * N, size=n_edges, replace=False)
        W_res[idx] = rng.standard_normal(n_edges)
        W_res = W_res.reshape(N, N)
        eigvals = np.linalg.eigvals(W_res)
        rho = np.max(np.abs(eigvals))
        if rho > 1e-8:
            W_res *= cfg.spectral_radius / rho
        self.W_res = W_res
        self.W_out: Optional[Array] = None
        self.state = np.zeros(N)
        self.last_input = np.zeros(n_input)

    def _step(self, u: Array) -> None:
        lr = self.cfg.leak_rate
        new = np.tanh(self.W_res @ self.state + self.W_in @ u)
        self.state = (1 - lr) * self.state + lr * new

    def train(self, data: Array, washout: int) -> None:
        n, d = data.shape
        N = self.cfg.N_reservoir
        self.state[:] = 0
        coll = np.zeros((n - washout - 1, N + d))
        for t in range(n - 1):
            self._step(data[t])
            if t >= washout:
                coll[t - washout, :N] = self.state
                coll[t - washout, N:] = data[t]
        targets = data[washout + 1:]
        A = coll.T @ coll + self.cfg.ridge_alpha * np.eye(N + d)
        self.W_out = np.linalg.solve(A, coll.T @ targets).T
        # Advance one step with final training point so state is correct for prediction
        self._step(data[-1])
        self.last_input = data[-1].copy()

    def predict(self, n_steps: int) -> Array:
        if self.W_out is None:
            raise RuntimeError("Train the ESN first.")
        d = self.W_out.shape[0]
        preds = np.zeros((n_steps, d))
        inp = self.last_input.copy()
        for t in range(n_steps):
            self._step(inp)
            out = self.W_out @ np.concatenate([self.state, inp])
            preds[t] = out
            inp = out
        return preds
